Upgrade to gRPC 0.11.

arielshaqed opened the pull request:
Mostly this adds "NULL" to various reserved locations. But completion
queue handling is now much simpler -- really only need to read a
single success value for the entire grpc_call_start_batch() call.

First part of fixing #21 

Currently this is unable to use the trusted CAs list in Ubuntu Trusty.

Commit:
Upgrade to gRPC 0.11.

Mostly this adds "NULL" to various reserved locations. But completion
queue handling is now much simpler -- really only need to read a
single success value for the entire grpc_call_start_batch() call.

Currently this is unable to use the trusted CAs list in Ubuntu Trusty.

arielshaqed commented:
@yosih please review.

yosih commented on src/write_grpc.c:
```diff
@@ -397,42 +398,29 @@ static gpr_timespec get_deadline(cdtime_t delay)
-  grpc_event *ev;
+  grpc_event ev;
 
   while (cb->write_accepted_counter < cb->write_counter) {
```
These 2 counters are unused

arielshaqed replied:
Removed them, and revamped the completion-queue wait and drain logic. PTAL at that, too (process_cq, drain_cq, and calls to them).

yosih commented on src/write_grpc.c:
```diff
@@ -582,18 +574,19 @@ static int do_flush_nolock(grpc_callback *cb, cdtime_t timeout) {
       &status, &status_details, &status_details_capacity);
   call = grpc_channel_create_call(
-      cb->channel,
+      cb->channel, NULL /* Parent call */, 0,
```
The extra parameters can go on separate lines

arielshaqed replied:
Done!

yosih commented on src/write_grpc.c:
```diff
@@ -810,80 +819,19 @@ static int write_data(const data_set_t *ds, const value_list_t *vl,
-
-  if (cred_client) grpc_credentials_release(cred_client);
-  if (cred_server) grpc_credentials_release(cred_server);
+  if (service_account_json_filename != NULL)
```
Environment variables are the only way to pass these parameters? Specifically for GRPC_GOOGLE_CREDENTIALS_ENV_VAR, if it's not set, will it go through the metadata server?

arielshaqed replied:
Yes, at least for this call :-(.

**tl;dr:** It is not clear how to compose an SSL server-side key and a JSON account client-side key. I discussed with Ido, he says to use the least amount of extra code. But AFAICT the library reads the environment variables _once_, and then reuses the values that it read. So if we have 2 `write_grpc` modules configured, `collectd` configuration is single-threaded and things should work.

**_Long...**_ The logic we have predates

yosih commented:
Thanks

Commit:
Code review fixes (1): `process_cq` logic was broken and relied on unused
counters; replaced with simpler wait and drain calls.

(With the previous gRPC version we could not do this because there were many
more events on the completion queue).

arielshaqed commented:
@yosih PTAL.

arielshaqed commented:
Thanks!
